File: api/anki_api.cpp
#include <iostream>
#include <string>
#include <cpprest/http_client.h>
#include <cpprest/json.h>

using namespace web;
using namespace web::http;
using namespace web::http::client;

class Client {
public:
  Client(const std::string& address, int port = 5555) 
    : port(port), address(address), base_url("http://" + address + ":" + std::to_string(port) + "/anki_api"), token("") {}

  pplx::task<std::string> create_car(const std::string& car) {
    http_client client(base_url);
    uri_builder builder("/create_car");
    builder.append_query("address", car);

    return client.request(methods::POST, builder.to_string())
      .then([](http_response response) {
        return response.extract_string();
      })
      .then([this](const std::string& token) {
        this->token = token;
        return token;
      });
  }

  pplx::task<std::string> set_speed(int speed, int accel = 1000) {
    http_client client(base_url);
    uri_builder builder("/set_speed");
    builder.append_query("token", token);
    builder.append_query("speed", speed);
    builder.append_query("accel", accel);

    return client.request(methods::POST, builder.to_string())
      .then([](http_response response) {
        return response.extract_string();
      });
  }

  pplx::task<std::string> left_lane(double lanes = 1.0) {
    http_client client(base_url);
    uri_builder builder("/left_lane");
    builder.append_query("token", token);
    builder.append_query("lanes", lanes);

    return client.request(methods::POST, builder.to_string())
      .then([](http_response response) {
        return response.extract_string();
      });
  }

  pplx::task<std::string> right_lane(double lanes = 1.0) {
    http_client client(base_url);
    uri_builder builder("/right_lane");
    builder.append_query("token", token);
    builder.append_query("lanes", lanes);

    return client.request(methods::POST, builder.to_string())
      .then([](http_response response) {
        return response.extract_string();
      });
  }

  pplx::task<std::vector<std::string>> get_cars(bool active = false) {
    http_client client(base_url);
    uri_builder builder("/get_cars");
    json::value body;
    body["active"] = json::value::boolean(active);

    return client.request(methods::POST, builder.to_string(), body.serialize(), "application/json")
      .then([](http_response response) {
        return response.extract_json();
      })
      .then([](json::value data) {
        std::vector<std::string> cars;
        for (const auto& car : data["cars"].as_array()) {
          cars.push_back(car.as_string());
        }
        return cars;
      });
  }

private:
  int port;
  std::string address;
  std::string base_url;
  std::string token;
};

File: api/anki_api.java
import java.io.IOException;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.net.http.HttpHeaders;
import java.util.concurrent.CompletableFuture;

public class Client {
    private int port;
    private String address;
    private String base_url;
    private String token;

    public Client(String address, int port) {
        this.port = port;
        this.address = address;
        this.base_url = "http://" + this.address + ":" + this.port + "/anki_api";
        this.token = "";
    }

    public CompletableFuture<String> create_car(String car) throws IOException, InterruptedException {
        HttpClient client = HttpClient.newHttpClient();
        HttpRequest request = HttpRequest.newBuilder()
                .uri(URI.create(base_url + "/create_car?address=" + car))
                .build();

        return client.sendAsync(request, HttpResponse.BodyHandlers.ofString())
                .thenApply(HttpResponse::body)
                .thenApply(token -> {
                    this.token = token;
                    return token;
                });
    }

    public CompletableFuture<String> set_speed(int speed, int accel) throws IOException, InterruptedException {
        HttpClient client = HttpClient.newHttpClient();
        HttpRequest request = HttpRequest.newBuilder()
                .uri(URI.create(base_url + "/set_speed?token=" + token + "&speed=" + speed + "&accel=" + accel))
                .build();

        return client.sendAsync(request, HttpResponse.BodyHandlers.ofString())
                .thenApply(HttpResponse::body);
    }

    public CompletableFuture<String> left_lane(double lanes) throws IOException, InterruptedException {
        HttpClient client = HttpClient.newHttpClient();
        HttpRequest request = HttpRequest.newBuilder()
                .uri(URI.create(base_url + "/left_lane?token=" + token + "&lanes=" + lanes))
                .build();

        return client.sendAsync(request, HttpResponse.BodyHandlers.ofString())
                .thenApply(HttpResponse::body);
    }

    public CompletableFuture<String> right_lane(double lanes) throws IOException, InterruptedException {
        HttpClient client = HttpClient.newHttpClient();
        HttpRequest request = HttpRequest.newBuilder()
                .uri(URI.create(base_url + "/right_lane?token=" + token + "&lanes=" + lanes))
                .build();

        return client.sendAsync(request, HttpResponse.BodyHandlers.ofString())
                .thenApply(HttpResponse::body);
    }

    public CompletableFuture<String[]> get_cars(boolean active) throws IOException, InterruptedException {
        HttpClient client = HttpClient.newHttpClient();
        HttpRequest request = HttpRequest.newBuilder()
                .uri(URI.create(base_url + "/get_cars"))
                .header("Content-Type", "application/json")
                .POST(HttpRequest.BodyPublishers.ofString("{\"active\":" + active + "}"))
                .build();

        return client.sendAsync(request, HttpResponse.BodyHandlers.ofString())
                .thenApply(HttpResponse::body)
                .thenApply(response -> {
                    HttpHeaders headers = response.headers();
                    String[] cars = headers.map().get("cars").get(0).split(",");
                    return cars;
                });
    }
}
